libraries/freertos_plus/standard/pkcs11/utest/iot_pkcs11_utest.c

```diff
@@ -0,0 +1,769 @@
+#include <stdbool.h>
+
+/* PKCS #11 includes. */
+#include "iot_pkcs11_config.h"
+#include "iot_pkcs11.h"
+
+/* Test includes. */
+#include "unity.h"
+
+/* Mock includes. */
+#include "mock_pkcs11.h"
+#include "mock_portable.h"
+
+/* ============================  GLOBAL VARIABLES =========================== */
+static CK_FUNCTION_LIST prvP11FunctionList =
+{
+    { CRYPTOKI_VERSION_MAJOR, CRYPTOKI_VERSION_MINOR },
+    C_Initialize,
+    C_Finalize,
+    NULL, /*C_GetInfo */
+    C_GetFunctionList,
+    C_GetSlotList,
+    C_GetSlotInfo,
+    C_GetTokenInfo,
+    NULL, /*C_GetMechanismList*/
+    C_GetMechanismInfo,
+    C_InitToken,
+    NULL, /*C_InitPIN*/
+    NULL, /*C_SetPIN*/
+    C_OpenSession,
+    C_CloseSession,
+    NULL,    /*C_CloseAllSessions*/
+    NULL,    /*C_GetSessionInfo*/
+    NULL,    /*C_GetOperationState*/
+    NULL,    /*C_SetOperationState*/
+    C_Login, /*C_Login*/
+    NULL,    /*C_Logout*/
+    C_CreateObject,
+    NULL,    /*C_CopyObject*/
+    C_DestroyObject,
+    NULL,    /*C_GetObjectSize*/
+    C_GetAttributeValue,
+    NULL,    /*C_SetAttributeValue*/
+    C_FindObjectsInit,
+    C_FindObjects,
+    C_FindObjectsFinal,
+    NULL, /*C_EncryptInit*/
+    NULL, /*C_Encrypt*/
+    NULL, /*C_EncryptUpdate */
+    NULL, /*C_EncryptFinal*/
+    NULL, /*C_DecryptInit*/
+    NULL, /*C_Decrypt*/
+    NULL, /*C_DecryptUpdate*/
+    NULL, /*C_DecryptFinal*/
+    C_DigestInit,
+    NULL, /*C_Digest*/
+    C_DigestUpdate,
+    NULL, /* C_DigestKey*/
+    C_DigestFinal,
+    C_SignInit,
+    C_Sign,
+    NULL, /*C_SignUpdate*/
+    NULL, /*C_SignFinal*/
+    NULL, /*C_SignRecoverInit*/
+    NULL, /*C_SignRecover*/
+    C_VerifyInit,
+    C_Verify,
+    NULL, /*C_VerifyUpdate*/
+    NULL, /*C_VerifyFinal*/
+    NULL, /*C_VerifyRecoverInit*/
+    NULL, /*C_VerifyRecover*/
+    NULL, /*C_DigestEncryptUpdate*/
+    NULL, /*C_DecryptDigestUpdate*/
+    NULL, /*C_SignEncryptUpdate*/
+    NULL, /*C_DecryptVerifyUpdate*/
+    NULL, /*C_GenerateKey*/
+    C_GenerateKeyPair,
+    NULL, /*C_WrapKey*/
+    NULL, /*C_UnwrapKey*/
+    NULL, /*C_DeriveKey*/
+    NULL, /*C_SeedRandom*/
+    C_GenerateRandom,
+    NULL, /*C_GetFunctionStatus*/
+    NULL, /*C_CancelFunction*/
+    NULL  /*C_WaitForSlotEvent*/
+};
+
+static uint16_t usMallocFreeCalls = 0;
+
+/* ==========================  CALLBACK FUNCTIONS =========================== */
+
+/*!
+ * @brief Wrapper stub for malloc.
+ */
+void * pvPkcs11MallocCb( size_t size,
+                         int numCalls )
+{
+    usMallocFreeCalls++;
+    return ( void * ) malloc( size );
+}
+
+/*!
+ * @brief Wrapper stub for malloc that fails every two calls.
+ *
+ * This is specifically to get complete branch coverage, as some branches
+ * are only taken if a previous call to malloc was successful.
+ */
+void * pvPkcs11MallocCb2( size_t size,
+                          int numCalls )
+{
+    static uint32_t ulCalls = 1;
+
+    ulCalls++;
+
+    if( ulCalls % 2 )
+    {
+        return NULL;
```

Comment:
Our test code may not need to be MISRA-compliant, but in general I'd get in the habit of not writing this early `return` pattern, especially in such a short function where it doesn't really improve readability.

Reply:
Applied this change.